### PR TITLE
[JUJU-1099] Restore "Store unit CharmURL as a string reference"

### PR DESCRIPTION
**Pull request by manadart**
This is an effective reapplication of #13546. See that patch for the full description.

That patch was reverted under #13636, which this in-turn reverts.

## QA steps

See #13546.

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

File: apiserver/facades/agent/uniter/uniter.go [hunk 1]
```diff
@@ -692,12 +692,21 @@ func (u *UniterAPI) CharmURL(args params.Entities) (params.StringBoolResults, er
...
-				})
-				curl, ok := charmURLer.CharmURL()
-				if curl != nil {
-					result.Results[i].Result = curl.String()
+				var cURL *charm.URL
+				var ok bool
+
+				switch entity := unitOrApplication.(type) {
+				case *state.Application:
+					cURL, ok = entity.CharmURL()
```
Comment (hpidcock) [inline apiserver/facades/agent/uniter/uniter.go:700]: Weird, looks like we need more work to make this a common interface

**Review by hpidcock — APPROVED** [review]

**Comment by manadart** [comment]
$$merge$$

File: apiserver/facades/agent/uniter/uniter.go (hunk 1 above)
Comment (manadart) [reply apiserver/facades/agent/uniter/uniter.go:700]: Yes, this only does the work required for the unit case, but we should address the other docs that store a `*charm.URL`, at which time a lot of double handling can be eliminated.

Comment (hmlanigan) [reply apiserver/facades/agent/uniter/uniter.go:700]: The name of `ok` is misleading here. For the unit, it was referring to the charm url pointer, which has been appropriately changed to an error. For application, the bool is whether a charm has been forced or not.

This facade is shared between the unit and application part of the uniter api - on the unit side, the bool is ignored. For the uniter application, force is used by the RemoteStateWorker.

**Pull request merged**